lib/openssh/sshd.go

```diff
@@ -0,0 +1,210 @@
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/trace"
+)
+
+const sshdBinary = "sshd"
+
+const sshdConfigSectionModificationHeader = "# Created by 'teleport join openssh', do not edit"
+
+var (
+	// AgentlessSSHDConfigPath is the path to write teleport specific SSHD config options
+	AgentlessSSHDConfigPath = "/etc/teleport/sshd.conf"
+	AgentlessKeysDir        = "/etc/teleport/openssh"
+	// agentlessSSHDConfig is Include directive added to the sshd_config file
+	agentlessSSHDConfigInclude = "Include " + AgentlessSSHDConfigPath
+)
+
+const (
+	TeleportKey       = "teleport"
+	TeleportCert      = "teleport-cert.pub"
+	TeleportOpenSSHCA = "teleport_user_ca.pub"
+)
+
+type sshdBackendOperations interface {
+	restart() error
+	checkConfig(path string) error
+}
+
+type SSHd struct {
+	sshd sshdBackendOperations
+}
+
+func NewSSHD() SSHd {
+	return SSHd{
+		sshd: &sshdBackend{},
+	}
+}
+
+func (b *SSHd) UpdateConfig(u SSHdConfigUpdate, restart bool) error {
+	needsUpdate, err := checkSSHDConfigAlreadyUpdated(u.SSHdConfigPath, agentlessSSHDConfigInclude)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if needsUpdate {
+		if err := prefixSSHDConfig(u.SSHdConfigPath, agentlessSSHDConfigInclude); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	configUpdate := fmtSSHDConfigUpdate(
+		u.OpenSSHCAPath,
+		u.HostKeyPaths,
+		u.HostCertPaths,
+	)
+
+	sshdConfigTmp, err := os.CreateTemp("", "")
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	defer sshdConfigTmp.Close()
+	if _, err := sshdConfigTmp.Write([]byte(configUpdate)); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	if err := sshdConfigTmp.Sync(); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	if err := b.sshd.checkConfig(sshdConfigTmp.Name()); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := os.Rename(sshdConfigTmp.Name(), u.TeleportSSHDConfigPath); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	if restart {
+		b.sshd.restart()
+	}
+
+	return nil
+}
+
+func WriteKeys(keysdir string, id *auth.Identity, cas []types.CertAuthority) error {
+	if err := os.WriteFile(filepath.Join(keysdir, TeleportKey), id.KeyBytes, 0600); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	if err := os.WriteFile(filepath.Join(keysdir, TeleportCert), id.CertBytes, 0600); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	var caKeyBytes []byte
+	for _, ca := range cas {
+		for _, key := range ca.GetTrustedSSHKeyPairs() {
+			caKeyBytes = append(caKeyBytes, key.PublicKey...)
+		}
+	}
+
+	if err := os.WriteFile(filepath.Join(keysdir, TeleportOpenSSHCA), caKeyBytes, 0600); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	return nil
+}
+
+func TryCreateDefaultAgentlesKeysDir(agentlessKeysPath string) error {
+	_, err := os.Stat(agentlessKeysPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return trace.Wrap(os.MkdirAll(agentlessKeysPath, 0700))
+		}
+		return trace.Wrap(err)
+	}
+	return nil
```

Comment:
Isn't this just `os.MkdirAll(agentlessKeysPath, 0o700)` but more complicated?

(also the explicit octal `0o700` is preferrable to the implicit one)